sympy/core/expr.py

```diff
@@ -40,26 +40,67 @@ def _diff_wrt(self):
         temporarily converts the non-Symbol vars in Symbols when performing
         the differentiation.
 
-        Note, see the docstring of Derivative for how this should work
-        mathematically. In particular, note that expr.subs(yourclass, Symbol)
-        should be well-defined on a structural level, or this will lead to
-        inconsistent results.
+        Non-symbol _diff_wrt=True variables must also implement a
+        _eval_derivative_wrt method. See Expr._eval_derivative_wrt
+        for details.
 
         Examples
         ========
 
-        >>> from sympy import Expr
+        >>> from sympy import Expr, Dummy
         >>> e = Expr()
         >>> e._diff_wrt
         False
         >>> class MyClass(Expr):
         ...     _diff_wrt = True
         ...
+        ...     def _eval_derivative_wrt(self, expr, new_name):
+        ...         new_self = Dummy(new_name)
+        ...         new_expr = expr.subs(self, new_self)
+        ...         return (new_expr, new_self)
+        ...
         >>> (2*MyClass()).diff(MyClass())
         2
         """
         return False
 
+    def _eval_derivative_wrt(self, expr, new_name):
+        """Transform derivative wrt a non-symbol variable to one wrt a symbol.
+
+        When taking the derivative of an expression with respect to (wrt) a
+        non-symbol variable with _diff_wrt=True, then this method is used to
+        create a new expression and a corresponding symbol variable.
+
+        Returns a tuple of new expr and symbol, or None if the transformation
+        could not be performed. The default implementation returns a tuple of
+        the 'new_expr' and 'new_self' where 'new_self' is a Dummy variable and
```

Comment:
These names are very confusing. In the function parameters, `self` is the object and `expr` is the thing we are taking a derivative with respect to, but here `new_expr` represents the modified `self` and `new_self` represents the replacement for `expr`.